### PR TITLE
Added Tooltip for Chat Avatars

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

- We are adding `Tooltip` to Avatar in all the chat options. This is added to the LHN, Header and the IOUPreview.

### Fixed Issues

$ https://github.com/Expensify/App/issues/7059

### Tests
1. Tested for 1:1 chat
2. Tested for multiple participants chat
3. Tested the same for HeaderView
4. Tested for Announce Rooms
5. Tested for pending IOU Preview

- [x] Verify that no errors appear in the JS console

### QA Steps

**LHN and Chat**
1. Open the app to the home screen on web/desktop
2. Go to the LHN and hover over `Avatar`
3. You should see the Tooltip
4. Click on the chat, go to the header of the chat
5. Hover over the `Avatar`, and the tooltip should show up

**IOU Preview**
1. Open any chat where there are pending IOU transactions on web/desktop
2. Click on the IOU badge on the LHN
3. In the IOUPreview of the transaction list summary, hover on the `Avatar` pair
4. Should show up `Tooltip`

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

web-multiple-avatar-tooltip-3
web-multiple-avatar-tooltip-2
web-multiple-avatar-tooltip-1

#### Mobile Web

NA

#### Desktop

desktop-multiple-avatars-tooltip-5
desktop-multiple-avatars-tooltip-4
desktop-multiple-avatars-tooltip-3
desktop-multiple-avatars-tooltip-2
desktop-multiple-avatars-tooltip-1

#### iOS

NA

#### Android

NA

### CODE REVIEW

**Comment by mananjadhav** [comment]
@parasharrajat @mountiny PR raised and two questions:

1. Will we ever have a case where the Tooltip text that I am setting will be empty? I have few chat rooms where the tooltip is empty. I am guessing those are from previous testing where room creation had errors.
2. I am using `managerEmail` and `ownerEmail` in `IOUPreview`. Is that the right selection?

**Review by mountiny — COMMENTED** [review]
Nice job Manan!

>Will we ever have a case where the Tooltip text that I am setting will be empty? I have few chat rooms where the tooltip is empty. I am guessing those are from previous testing where room creation had errors.

That is a good question. I think it might be worth it to prevent showing the tooltip if the string is empty. We might have to add some boilerplate to the files, but I believe it would be worth it in case there is nothing to show in them for whatever reason. Thoughts @parasharrajat?

> I am using managerEmail and ownerEmail in IOUPreview. Is that the right selection?

That seems good to me.

**Comment by parasharrajat** [comment]
> Will we ever have a case where the Tooltip text that I am setting will be empty? I have few chat rooms where the tooltip is empty. I am guessing those are from previous testing where room creation had errors.

I think it would be better to skip the tooltip when there is no title. Maybe move this logic to Tooltip itself. No need to initialize when there is nothing to show.

**Comment by mananjadhav** [comment]
@parasharrajat PR updated with the change in Tooltip

**Comment by parasharrajat** [comment]
I have a few concerns: 
1. I think we should show separate tooltips for the Multiple Avatars. Each avatar will have its own tootltip.
2. Should we remove the tooltip from the IOU preview? I don't feel that it is needed. But if we keep it, the first point applies.
3. Avatar component covers multiple parts of the app. @mananjadhav Could you please list down the places where the change is applied? Then we can pick where should we disable that. One possible target to disable tooltip would be Workspace Invite page.

**Comment by mananjadhav** [comment]
I had intended to create this for chat related Avatar (MultipleAvatar) component. I'll add places where Avatar is used and accordingly take the call.

**Review by mountiny — COMMENTED** [review]
I am now while looking at the screenshots thinking, whether we would like to limit the width of the tooltip. 
Given this screenshot 
image

I assume that with more people in the group the tooltip could be taking an entire width of the screen. This could be weird UX if user is just moving their cursor around.

Would you say you share this worry @mananjadhav and @parasharrajat? We could limit the string and show `...` from certain lengths of the string.

Thoughts?

**Comment by mountiny** [comment]
Good call @parasharrajat.

**Comment by mountiny** [comment]
There is this issue [here](https://github.com/Expensify/App/issues/7481), would you be able to say you experience the same problems in this PR branch?

**Comment by parasharrajat** [comment]
@mountiny that would be a different issue and let's add that to https://github.com/Expensify/App/issues/7481.

**Comment by mountiny** [comment]
@parasharrajat I agree, was looking to see if you guys agree this could be a problem.

I would say it might be better to keep it one line, it might not show a lot of emails in the end, but putting it on next lines will just make a block of dense small text which might be hard to read.

**Comment by parasharrajat** [comment]
> I would say it might be better to keep it one line, it might not show a lot of emails in the end, but putting it on next lines will just make a block of dense small text which might be hard to read.

But it loses the purpose of the tooltip data in many locations. for example, Tooltip was added to the LHN `...` to see the rest of the users of the group that is not visible on the report title. So if we trim that it won't look good. 

but wrapping and creating a box before clipping would be sufficient to fit most of the data. Block tooltip does not look bad to me. I have seen that most of the apps have blocky tooltips.

We can raise this in Slack to get more opinions but main thing is that I agree this is a issue.

**Comment by mountiny** [comment]
As mentioned here, I have created a new issue https://github.com/Expensify/App/issues/7486 to handle the tooltip size. I will keep it as noon-external until we merge this PR.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
https://github.com/Expensify/App/pull/7459#issuecomment-1025481203

**Comment by mountiny** [comment]
> 1. I think we should show separate tooltips for the Multiple Avatars. Each avatar will have its own tootltip.
> 2. Should we remove the tooltip from the IOU preview? I don't feel that it is needed. But if we keep it, the first point applies.

Sorry if it was not clear. I agree with your points, so lets try to do it, if it is not too complicated @mananjadhav .

**Comment by mananjadhav** [comment]
Ohh I thought we didn’t finalise. For the point one when we have more than two we show Avatar + 4. So I’ll show the name for the first one and for the “4” I’ll show comma separated values ? 

@parasharrajat @mountiny  ^

**Comment by mountiny** [comment]
@mananjadhav Correct! that is how I would imagine it would work. Thanks for double checking.

**Comment by mananjadhav** [comment]
@parasharrajat  @mountiny I've updated the PR.

File: src/components/MultipleAvatars.js
Comment (mountiny) [inline src/components/MultipleAvatars.js:69]: I think a short comment here might be useful for readability.

**Review by mountiny — COMMENTED** [review]
Great job so far Manan, thank you for working on this. I gave it just a quick look now and left one feedback. I think a comment might be useful for that tooltip content logic.

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [inline src/components/MultipleAvatars.js:60]: this change? It does not seem required but let's do some testing on this and see why it was added [here](https://github.com/Expensify/App/pull/3129/files)

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [inline src/components/MultipleAvatars.js:77]:
```suggestion
                <Tooltip text={props.avatarTooltips[0]} absolute>
```

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [inline src/components/MultipleAvatars.js:87]:
```suggestion
                        <Tooltip text={props.avatarTooltips[1]} absolute>
```

File: src/components/Tooltip/index.js
Comment (parasharrajat) [inline src/components/Tooltip/index.js:166]: Please add a small comment...

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [inline src/components/MultipleAvatars.js:94]:
```suggestion
                        <Tooltip text={tooltip.slice(1).join(', ')} absolute>
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/MultipleAvatars.js
Comment (mananjadhav) [reply src/components/MultipleAvatars.js:69]: I guess with the new comments, this won't be required now.

File: src/components/MultipleAvatars.js
Comment (mananjadhav) [reply src/components/MultipleAvatars.js:60]: If I don't remove `pointerEvents="none"`, the Tooltip doesn't show up.

**Comment by mananjadhav** [comment]
@parasharrajat @mountiny  PR updated with the relevant changes.

File: src/components/MultipleAvatars.js
Comment (mountiny) [reply src/components/MultipleAvatars.js:60]: No with this prop being removed though, can you replicate this with 3D touch: https://github.com/Expensify/App/issues/3019? 

I mean 3D touch is no longer supported on the newest iphones so if this is the main concern, _maybe_ we could not worry about it. But better to not introduce new regressions.

**Review by mountiny — COMMENTED** [review]
Thanks for the changes, looks very good now!

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [reply src/components/MultipleAvatars.js:60]: Yup. I agree with @mountiny. @mananjadhav Could you please do some testing for #3019. If it is causing issues then we can talk about alternatives.

**Review by parasharrajat — COMMENTED** [review]
Changes look convincing. Testing shortly... on my list 3/5.

File: src/components/MultipleAvatars.js
Comment (mananjadhav) [reply src/components/MultipleAvatars.js:60]: Thanks for pointing out the related GH issue. I'll test it out.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Disable tooltips from Workspace Member page and invite member page.

and all other places where you don't see tooltip on the report title/chat name/user name.

**Comment by mananjadhav** [comment]
> Disable tooltips from Workspace Member page and invite member page.

@parasharrajat  Are we saying we won't show tooltip at these places? Now, these use `OptionRow`. I'll  have to add one more prop to `OptionRow` say `hideAvatarTooltip`

> and all other places where you don't see tooltip on the report title/chat name/user name.

I didn't understand this.

**Comment by parasharrajat** [comment]
> Are we saying we won't show tooltips at these places? 

Yeah. I know it makes it complicated but those places do not have a tooltip on the report title so I think we can keep this consistent.

> and all other places where you don't see tooltip on the report title/chat name/user name.

Like NewChat page. If you hover over the title/name in the NewChat page list. It does not show tooltip. 


If you have second opinion let's open a poll about it on Slack.

**Comment by mananjadhav** [comment]
> If you have second opinion let's open a poll about it on Slack.

Yeah let’s do that? More than the second opinion, don’t want to undo some code or probably get another issue reported in future to enable it again. 

I’ll post it on slack, if you’re okay with checking it once

**Comment by mananjadhav** [comment]
Posted this on slack https://expensify.slack.com/archives/C01GTK53T8Q/p1644361109816329 here, lets wait for the responses?

**Comment by mountiny** [comment]
@mananjadhav We did not get much input in the Slack thread but just going by the screenshot, the email is already shown in the view with in grey subtitle colour. That makes me feel like we should not show the tooltip as it is. just cluttering the interface with redundant information, which is already shown to the user. Would you be able to work in that change?

**Comment by mananjadhav** [comment]
Give me 1 day I'll push an update.

**Review by mountiny — COMMENTED** [review]
Sure, no worries! Thank you for the message!

**Comment by mananjadhav** [comment]
Was a bit unwell, will push an update tonight.

**Comment by mountiny** [comment]
No worries, I hope you feeling better, Manan!

**Comment by mananjadhav** [comment]
@parasharrajat @mountiny I am using `showTitleTooltip` of `OptionRow` to hide the avatar tooltip as well. I was able to check this in Workspace Members Page, Invite Page, and IOU User selection. This should work, let me know if you think there are any challenges.

Secondly, I've added Tooltip to individual chat row avatars because their title counterparts have the tooltip enabled. This is to ensure the consistency we were aiming by hiding tooltip in other areas.

Please let me know if there are any other changes.

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [reply src/components/MultipleAvatars.js:60]: Any update here @mananjadhav

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [inline src/components/MultipleAvatars.js:104]:
```suggestion
```

File: src/components/MultipleAvatars.js
Comment (mananjadhav) [reply src/components/MultipleAvatars.js:60]: Did hide the tooltip for the workspace section. Missed this one to test. I'll check by tonight and update.

File: src/pages/home/sidebar/OptionRow.js
Comment (parasharrajat) [inline src/pages/home/sidebar/OptionRow.js:132]:
```suggestion
    const avatarTooltips = props.showTitleTooltip ? _.pluck(displayNamesWithTooltips, 'tooltip') : undefined;
```

File: src/pages/home/sidebar/OptionRow.js
Comment (parasharrajat) [inline src/pages/home/sidebar/OptionRow.js:180]:
```suggestion
                                        avatarTooltips={avatarTooltips}
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Review by parasharrajat — CHANGES_REQUESTED** [review]
There should be no tooltips for the Room. But currently, it is shown on the avatar.
image

**Review by mountiny — COMMENTED** [review]
Great job so far. thank you for sticking with this issue. I am not sure how much time you have spent on this PR, but if you feel like the scope has grown over the original proposal, feel free to follow [these instructions](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#payment-for-contributions) and comment on the original issue regarding compensation.

File: src/components/MultipleAvatars.js
Comment (mananjadhav) [reply src/components/MultipleAvatars.js:60]: I tested this and nothing broke. But I am not too sure about reproducing the issue. I have checked this on iPhone 8, which says has 3D Touch-enabled, but I couldn't find this setting.

Is there any other way to test it?

Screenshot 2022-02-20 at 11 21 55 PM
Screenshot 2022-02-20 at 11 13 20 PM

Comment (mountiny) [reply src/components/MultipleAvatars.js:60]: I guess in general Long press on newer iPhones too since 3D touch has been deprecated and they use Long press instead. I believe it might have similar behaviour.

Comment (mananjadhav) [reply src/components/MultipleAvatars.js:60]: I tried Long press on iPhone 12 and iPhone 8, for which I've uploaded the screenshots above.

Comment (mountiny) [reply src/components/MultipleAvatars.js:60]: However, I think the long press will be fine and will do what you have shown on the screens in the thread. So I guess this is fine.

**Comment by mananjadhav** [comment]
> There should be no tooltips for the Room. But currently, it is shown on the avatar.

@parasharrajat @mountiny It seems for Room handling we'll have to update `BaseOptionsList`

https://github.com/Expensify/App/blob/c873776c3c3beaed86ee6ce986bb7a083b43a228/src/components/OptionsList/BaseOptionsList.js#L101-L106

I am thinking of setting it as: `showTitleTooltip={_.isBoolean(item.showTitleTooltip)  ? item.showTitleTooltip:  this.props.showTitleTooltip}`

Is that fine?

File: src/pages/home/sidebar/OptionRow.js
Comment (parasharrajat) [inline src/pages/home/sidebar/OptionRow.js:133]:
```suggestion
    const avatarTooltips = props.showTitleTooltip && !props.option.isChatRoom && !props.option.isArchivedRoom ? _.pluck(displayNamesWithTooltips, 'tooltip') : undefined;

```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/home/sidebar/OptionRow.js
Comment (mananjadhav) [reply src/pages/home/sidebar/OptionRow.js]: Thanks, for this. I totally missed `props.option` object.

File: src/components/MultipleAvatars.js
Comment (parasharrajat) [reply src/components/MultipleAvatars.js:60]: Thanks, @mananjadhav for looking into this. Let's add QA steps to test https://github.com/Expensify/App/issues/3019. Do mention to test https://github.com/Expensify/App/issues/3019.

**Review by parasharrajat — APPROVED** [review]
LGTM. 

cc: @mountiny 
:ribbon: :eyes: :ribbon:  C+ reviewed

**Review by mountiny — APPROVED** [review]
@mananjadhav Amazing job on this one Manan, very glad you were able to stick through the iterations here. Thank you very much and thanks @parasharrajat for great job here too!

**Comment by Stutikuls** [comment]
**Issue 1 -**

**Title**-[Medium] Chrome: Keyboard : Tooltip is not accessible using keyboard for 'Setting' control.
**Actual** - Tooltip is appearing using mouse but it's not appearing using keyboard for 'Setting' control.
           **Note -** Same issue repro on the whole expensify page.
**Expected** - Tooltip should accessible/accessible using keyboard for 'Setting' control.
**WCAG failure** - 2.1.1
**Reproducible in staging?**: - Yes
**Version Number:** - v1.1.40-2
**Platforms** - Web(Chrome + Jaws), Desktop(MAC)

**Note** - Due to installation of build not able to check on Android

https://user-images.githubusercontent.com/97600950/157241211-d1d9ee61-1607-42d3-8f83-95a4063645ee.mp4

**Pull request merged**
